test/google/gax/api_call/retry_raise_test.rb

```diff
@@ -0,0 +1,136 @@
+# Copyright 2019, Google Inc.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+#     * Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above
+# copyright notice, this list of conditions and the following disclaimer
+# in the documentation and/or other materials provided with the
+# distribution.
+#     * Neither the name of Google Inc. nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# 'AS IS' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+require 'test_helper'
+
+class ApiCallRetryTest < Minitest::Test
+  def test_no_retry_without_codes
+    call_count = 0
+    api_meth_stub = proc do
+      call_count += 1
+      raise GRPC::BadStatus.new(FAKE_STATUS_CODE_1, 'unknown')
+    end
+
+    api_call = Google::Gax::ApiCall.new(api_meth_stub)
+
+    assert_raises Google::Gax::GaxError do
+      api_call.call Object.new
+    end
+    assert_equal(1, call_count)
+  end
+
+  def test_aborts_retries
+    api_meth_stub = proc do
+      raise GRPC::BadStatus.new(FAKE_STATUS_CODE_1, 'unknown')
+    end
+    api_call = Google::Gax::ApiCall.new(api_meth_stub)
+
+    exc = assert_raises Google::Gax::GaxError do
+      api_call.call Object.new
+      raise 'This will never be reached'
+    end
+    assert_kind_of(GRPC::BadStatus, exc.cause)
+  end
+
+  def test_times_out
+    to_attempt = 5
+    call_count = 0
+    deadline_arg = nil
+
+    api_meth_stub = proc do |deadline: nil, **_kwargs|
+      deadline_arg = deadline
+      call_count += 1
+      raise GRPC::BadStatus.new(FAKE_STATUS_CODE_1, 'unknown')
+    end
+
+    api_call = Google::Gax::ApiCall.new(api_meth_stub)
+    options = Google::Gax::CallOptions.new(
+      retry_policy: { retry_codes: [14, 101] }
```

Comment:
This is ok for tests, but it might help to put a note here if these codes are actual codes (i.e. what constant to they correspond to) or are out of range.

Reply:
These tests are ports of the original tests, which were all written using fake errors and fake `code` values. I am not a huge fan of them either.

Do you approve of the overall approach? I don't want to spend time writing better tests if the approach is flawed.

Reply:
Ah, ok. Overall approach is great - thanks for doing it! 

I do think we should address any omissions in the tests now though.

Reply:
Okay, tests have been updated to use real GRPC error codes, and to only retry GRPC::BadStatus errors.